### PR TITLE
changes: sleep time, constants, etc.

### PR DESCRIPTION
**Pull request by m3da9ti**

### CODE REVIEW

File: e4-osc-bridge.py [hunk 1]
```diff
@@ -95,17 +105,18 @@ def start_replay(replay_file, osc_ip, osc_port):
...
         line = line.strip()
         event_time, event_type, *sample = line.split(",")
         event_time = float(event_time)
         sample = [float(x) for x in sample]
 
-        time.sleep(event_time - last_time)
-        last_time = event_time
+        # print(event_time, last_time)
+        time.sleep(abs(event_time - last_time.get(event_type,0)))
+        last_time[event_type] = event_time
```
Comment (fdb) [inline e4-osc-bridge.py:117]: I'm not so sure about this part... All the events should ideally be sorted in the file per date, so sleeping should happen until the next event, not the next event of that type.

**Comment by m3da9ti** [comment]
events were not sorted by time stamp which caused time.sleeo to give an error of negative value so this was to remedy that.

Sent from Yahoo Mail on Android 
 
  On Sun., 18 Sep. 2022 at 9:12 p.m., Frederik De ***@***.***> wrote:   


@fdb commented on this pull request.

In e4-osc-bridge.py:
>      for line in log_file:
         line = line.strip()
         event_time, event_type, *sample = line.split(",")
         event_time = float(event_time)
         sample = [float(x) for x in sample]
 
-        time.sleep(event_time - last_time)
-        last_time = event_time
+        # print(event_time, last_time)
+        time.sleep(abs(event_time - last_time.get(event_type,0)))
+        last_time[event_type] = event_time

I'm not so sure about this part... All the events should ideally be sorted in the file per date, so sleeping should happen until the next event, not the next event of that type.

**Comment by fdb** [comment]
@m3da9ti 
The proposed fix, to categorise events by type and sleep based on those, creates sync errors in the replay, however. 

Imagine the following scenario (values are omitted):

| time | event |
| ----- | ----- |
| 1.1   | acc  |
| 1.2  | temp |   
| 1.3 | temp | 
| 1.2 | acc |

In this case, the last two events appear in the wrong order, but the problems already happen earlier:

1. The first `acc` event is loaded. The script sleeps for 1.1 seconds and fires the event.
2. The second `temp` event is loaded. This is a different type, so the system looks for the event of that type. It can't find it, so it sleeps (incorrectly) for 1.2 seconds, not 0.1 seconds as expected. The system is now more than a second out of sync.
3. The third `temp` event is loaded. This is the same type, so the system will sleep (correctly) for 0.1 seconds.
4. The fourth `acc` event is loaded. The system looks for the last time an `acc` event was fired, which was at 1.1, so the system will sleep (incorrectly) for 0.1 seconds.

Clearly this is not the desired outcome, as events will quickly go out of sync.

I think the correct solution is to sort the file by timestamp when loading it. I'll update the code and test it with a dummy log file.

**Comment by fdb** [comment]
e73cc19b1b4a9f434203caf82263d503b45a1dc5 — this should be the correct implementation. I order the timestamps and sleep accordingly.

**Comment by m3da9ti** [comment]
sounds good!!

Sent from Yahoo Mail on Android 
 
  On Sun., 18 Sep. 2022 at 11:08 p.m., Frederik De ***@***.***> wrote:   


@m3da9ti
The proposed fix, to categorise events by type and sleep based on those, creates sync errors in the replay, however.

Imagine the following scenario (values are omitted):

| time | event |
| 1.1 | acc |
| 1.2 | temp |
| 1.3 | temp |
| 1.2 | acc |


In this case, the last two events appear in the wrong order, but the problems already happen earlier:
   
   - The first acc event is loaded. The script sleeps for 1.1 seconds and fires the event.
   - The second temp event is loaded. This is a different type, so the system looks for the event of that type. It can't find it, so it sleeps (incorrectly) for 1.2 seconds, not 0.1 seconds as expected. The system is now more than a second out of sync.
   - The third temp event is loaded. This is the same type, so the system will sleep (correctly) for 0.1 seconds.
   - The fourth acc event is loaded. The system looks for the last time an acc event was fired, which was at 1.1, so the system will sleep (incorrectly) for 0.1 seconds.

Clearly this is not the desired outcome, as events will quickly go out of sync.

I think the correct solution is to sort the file by timestamp when loading it. I'll update the code and test it with a dummy log file.

**Pull request merged**